### ISSUE TITLE
Nextjs SSR has extreme TTFB on firebase cloud functions +30sec!

### ISSUE THREAD

**Issue by alexvandervegt** [zeit/next.js#10491]
# Bug report
When I enter an URL the first time, the TTFB is 30sec+. After that when I click links everything is very fast as expected.

I suspect the `app.getRequestHandler()` since this is the last step on the server before the Page.getInitialProps (see log in screen) is called.

## To Reproduce

If needed I can email an staging link, you can contact me on alexvdvegt |at| vegtmedia.com

**_app.js**

```
import React from 'react';
import {ThemeProvider} from '@material-ui/styles';
import CssBaseline from '@material-ui/core/CssBaseline';
import theme from '../theme';
import SnackbarProvider from 'providers/SnackbarProvider';
import GlobalProvider from 'providers/GlobalProvider';
import Layout from 'components/Layout';

export default function ({Component, pageProps}) {
    return (
        <React.Fragment>
            <ThemeProvider theme={theme}>
                <SnackbarProvider>
                    <GlobalProvider>
                        <Layout>
                            <CssBaseline />
                            <Component {...pageProps} />
                        </Layout>
                    </GlobalProvider>
                </SnackbarProvider>
            </ThemeProvider>
        </React.Fragment>
    );
}
```

**_document.js**

```
import React from 'react';
import Document, {Head, Main, NextScript} from 'next/document';
import {ServerStyleSheets} from '@material-ui/styles';
import theme from '../theme';

class MyDocument extends Document {
    render() {
        return (
            <html lang="nl">
            <Head>
                <link
                    rel="stylesheet"
                    href="https://fonts.googleapis.com/css?family=Overlock:900&display=swap"
                />
                <meta charSet="utf-8" />
                <meta
                    name="viewport"
                    content="width=device-width, initial-scale=1, shrink-to-fit=no"
                />
                <meta
                    name="apple-mobile-web-app-capable"
                    content="yes"
                />
                <meta
                    name="apple-mobile-web-app-title"
                    content="Bemmel & Kroon"
                />
                <meta
                    name="apple-mobile-web-app-status-bar-style"
                    content="black"
                />
                <meta
                    name="theme-color"
                    content={theme.palette.primary.main}
                />
                <meta
                    name="format-detection"
                    content="telephone=no"
                />
                <meta
                    name="HandheldFriendly"
                    content="true"
                />
                <meta
                    name="MobileOptimized"
                    content="375"
                />
                <meta
                    name="mobile-web-app-capable"
                    content="yes"
                />
                <link
                    rel="manifest"
                    href="/manifest.json"
                />
                <link
                    rel="icon"
                    type="image/png"
                    sizes="32x32"
                    href="/icons/favicon-32x32.png"
                />
                <link
                    rel="icon"
                    type="image/png"
                    sizes="16x16"
                    href="/icons/favicon-16x16.png"
                />
                <link
                    rel="apple-touch-icon"
                    sizes="180x180"
                    href="/icons/apple-icon-180.png"
                />
                <link
                    rel="apple-touch-icon"
                    sizes="167x167"
                    href="/icons/apple-icon-167.png"
                />
                <link
                    rel="apple-touch-icon"
                    sizes="152x152"
                    href="/icons/apple-icon-152.png"
                />
                <link
                    rel="apple-touch-icon"
                    sizes="120x120"
                    href="/icons/apple-icon-120.png"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-2048-2732.png"
                    media="(device-width: 1024px) and (device-height: 1366px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-2732-2048.png"
                    media="(device-width: 1024px) and (device-height: 1366px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-1668-2388.png"
                    media="(device-width: 834px) and (device-height: 1194px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-2388-1668.png"
                    media="(device-width: 834px) and (device-height: 1194px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-1668-2224.png"
                    media="(device-width: 834px) and (device-height: 1112px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-2224-1668.png"
                    media="(device-width: 834px) and (device-height: 1112px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-1536-2048.png"
                    media="(device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-2048-1536.png"
                    media="(device-width: 768px) and (device-height: 1024px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-1242-2688.png"
                    media="(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-2688-1242.png"
                    media="(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-1125-2436.png"
                    media="(device-width: 375px) and (device-height: 812px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-2436-1125.png"
                    media="(device-width: 375px) and (device-height: 812px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-828-1792.png"
                    media="(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-1792-828.png"
                    media="(device-width: 414px) and (device-height: 896px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-1242-2208.png"
                    media="(device-width: 414px) and (device-height: 736px) and (-webkit-device-pixel-ratio: 3) and (orientation: portrait)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-2208-1242.png"
                    media="(device-width: 414px) and (device-height: 736px) and (-webkit-device-pixel-ratio: 3) and (orientation: landscape)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-750-1334.png"
                    media="(device-width: 375px) and (device-height: 667px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-1334-750.png"
                    media="(device-width: 375px) and (device-height: 667px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-640-1136.png"
                    media="(device-width: 320px) and (device-height: 568px) and (-webkit-device-pixel-ratio: 2) and (orientation: portrait)"
                />
                <link
                    rel="apple-touch-startup-image"
                    href="/icons/apple-splash-1136-640.png"
                    media="(device-width: 320px) and (device-height: 568px) and (-webkit-device-pixel-ratio: 2) and (orientation: landscape)"
                />
            </Head>
            <body>
            <Main />
            <NextScript />
            </body>
            </html>
        );
    }
}

MyDocument.getInitialProps = async (ctx) => {
    // Resolution order
    //
    // On the server:
    // 1. app.getInitialProps
    // 2. page.getInitialProps
    // 3. document.getInitialProps
    // 4. app.render
    // 5. page.render
    // 6. document.render
    //
    // On the server with error:
    // 1. document.getInitialProps
    // 2. app.render
    // 3. page.render
    // 4. document.render
    //
    // On the client
    // 1. app.getInitialProps
    // 2. page.getInitialProps
    // 3. app.render
    // 4. page.render

    // Render app and page and get the context of the page with collected side effects.
    const sheets = new ServerStyleSheets();
    const originalRenderPage = ctx.renderPage;

    ctx.renderPage = () =>
        originalRenderPage({
            enhanceApp: App => props => sheets.collect(<App {...props} />),
        });

    const initialProps = await Document.getInitialProps(ctx);

    return {
        ...initialProps,
        // Styles fragment is rendered after the app and page rendering finish.
        styles: [...React.Children.toArray(initialProps.styles), sheets.getStyleElement()],
    };
};

export default MyDocument;
```

**page.js => getInitialProps**

```
Page.getInitialProps = async ({res, query}) => {
    let data = false;

    console.log(`[TIME][Page.getInitialProps]: ${(new Date()).getHours()}:${(new Date()).getMinutes()}:${(new Date()).getSeconds()}:${(new Date()).getMilliseconds()}`);

    try {
        let rewrite = res?.locals?.rewrite;

        if (!rewrite) {
            let path = `/${(query.rewrite) ? query.rewrite.join('/') : ''}`;

            let rewrites = await firestore.collection(`rewrites`).where('path', '==', path).get();

            if (rewrites.docs.length > 0) {
                rewrites.forEach((doc) => {
                    rewrite = {
                        id: doc.id,
                        ...doc.data(),
                    };
                });
            }
        }

        if (rewrite) {
            let hit = await firestore.doc(`${rewrite.type}s/${rewrite.id}`).get();

            data = {
                ...hit.data(),
                id: hit.id,
                path: rewrite.path,
                type: rewrite.type,
                blocks: [],
            };
        }
    } catch (error) {
        console.error(error);
    }

    return {data};
};

export default Page;
```

**server.js**

```
import * as functions from 'firebase-functions';
import * as admin from 'firebase-admin';
import next from 'next';
import express from 'express';
import path from 'path';
import rewriteHandler from './rewriteHandler';

admin.initializeApp();

const app = next({
    dev: false,
    conf: {
        distDir: `${path.relative(process.cwd(), __dirname)}/next`,
    },
});
const handle = app.getRequestHandler();
const server = express();

server.disable('x-powered-by');
server.set('trust proxy', 1);
server.use(rewriteHandler);

server.get('*', async (request, response) => {
    // Proper 404 handling
    if (response.locals?.error === 404) {
        response.status(404);

        return app.render(request, response, '/_error', {
            ...request.query,
            ...request.params,
            statusCode: 404,
        });
    }

    console.log(`[TIME][server]: ${(new Date()).getHours()}:${(new Date()).getMinutes()}:${(new Date()).getSeconds()}:${(new Date()).getMilliseconds()}`);

    return handle(request, response);
});

exports.next = functions.https.onRequest(async (request, response) => {
    await app.prepare();

    return server(request, response);
});
```

## Expected behavior

Normal TTFB like on firebase simulation environment on local machine.

## Screenshots
<img width="1618" alt="Screenshot 2020-02-11 at 10 25 38" src="https://user-images.githubusercontent.com/3672692/74224951-13522500-4cba-11ea-9454-57f82734232a.png">
<img width="381" alt="Screenshot 2020-02-11 at 10 26 56" src="https://user-images.githubusercontent.com/3672692/74224961-177e4280-4cba-11ea-9a7b-55308a228402.png">

## Additional context
I think it is related to this issue:
https://github.com/zeit/next.js/issues/5882

But there is no solution provided in there and I tried to contact @hrdymchl but I have no reply yet.

I hope somebody can see what is going wrong.

**Comment by alexvandervegt** [comment]
This is solved, it was in one of my components

**Comment by shovelend** [comment]
@alexvandervegt We're seeing similar issues on our end. What was wrong in one of your components?

**Comment by alexvandervegt** [comment]
I moved my context providers from the _app.js to the particular pages which actually use the provider instead of all in _app

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
